src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,214 @@
+    /// match scrutinee, its lifetime lasts until the end of the match block, which may be
+    /// surprising. It can also lead to deadlocks if the match scrutinee uses a function call that
+    /// returns a MutexGuard and then tries to lock again in one of the match arms, because the
+    /// MutexGuard will not be dropped until the end of the match block and thus will not unlock.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     match mutex.lock().unwrap().foo() {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar(); // Deadlock!
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some(expr_span) = has_significant_drop_in_scrutinee(cx, expr) {
+            span_lint_and_help(
+                cx,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                expr_span,
+                "The lifetimes of temporaries in a match scrutinee last to the end of the \
+                match expressions. One or more of the temporaries in this match scrutinee have a \
+                drop with a side-effect (such as unlocking a mutex). Consider calling the scrutinee \
+                outside of the match expression and using the result instead.",
+                None,
+                format!(
+                    "Try this:\nlet value = {};\nmatch value {{",
+                    snippet(cx, expr_span, "..")
+                )
+                .as_str(),
```

Comment:
```suggestion
                &format!(
                    "Try this:\nlet value = {};\nmatch value {{",
                    snippet(cx, expr_span, "..")
                ),
```

This should also work.

---

Besides that, this should be a `span_suggestion()`. With that, `rustfix` can automatically apply the suggestion. Also here the text should start lowercase and don't have punctuation at the end.

Reply:
Thank you for the suggestion! I ended up struggling a bit with how to use the span/suggestions with this multi-line suggestion, but I think I have it working.